### ISSUE TITLE
[REVIEW]: CLAIRE: Constrained Large Deformation Diffeomorphic Image Registration on Parallel Computing Architectures

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3038]
**Submitting author:** @andreasmang (<a href="http://orcid.org/0000-0003-4605-3290">Andreas Mang</a>)
**Repository:** <a href="https://github.com/andreasmang/claire" target ="_blank">https://github.com/andreasmang/claire</a>
**Version:** v0.08
**Editor:** @mjsottile
**Reviewers:** @tjasaki
**Archive:** <a href="https://doi.org/10.5281/zenodo.4837427" target="_blank">10.5281/zenodo.4837427</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/d4af0a15946fd2161935be018820243b/status.svg)](https://joss.theoj.org/papers/d4af0a15946fd2161935be018820243b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/d4af0a15946fd2161935be018820243b"><img src="https://joss.theoj.org/papers/d4af0a15946fd2161935be018820243b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/d4af0a15946fd2161935be018820243b/status.svg)](https://joss.theoj.org/papers/d4af0a15946fd2161935be018820243b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/andreasmang/claire" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@tjasaki & @NMontanaBrown, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mjsottile know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/andreasmang/claire">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@andreasmang) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/andreasmang/claire">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@andreasmang) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @tjasaki, @NMontanaBrown it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.79 s (171.5 files/s, 68268.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C++                             44           8821           6972          27289
C/C++ Header                    48           1464           1122           3491
Bourne Shell                    14            330            317           1504
CMake                           14            165            418            989
Markdown                         9            179              0            480
TeX                              1             33              4            235
make                             3             40              7            180
Python                           1             17              7             60
YAML                             2              0              0              5
-------------------------------------------------------------------------------
SUM:                           136          11049           8847          34233
-------------------------------------------------------------------------------


Statistical information for the repository '181063aa74ae26a70dd5afc3' was
gathered on 2021/02/13.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Amir Gholami                     1             1              0            0.00
Andreas Mang                   588         94954          44608           91.03
James Herring                    5           765             91            0.56
MalteBrunn                       2           228           6005            4.07
Naveen Himthani                 21          4087           1328            3.53
brunnme                          1             6              0            0.00
naveenaero                       4          1241              6            0.81

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Amir Gholami                  1          100.0         33.8                0.00
Andreas Mang              46628           49.1         38.5               22.07
James Herring               702           91.8         20.0               16.10
MalteBrunn                  234          102.6         16.0                2.14
Naveen Himthani              13            0.3         14.5               15.38
naveenaero                 1673          134.8          9.6                6.81
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03038/joss.03038/10.21105.joss.03038.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03038/joss.03038/10.21105.joss.03038.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1038/nprot.2014.123 may be a valid DOI for title: Advanced CLARITY for rapid and high-resolution imaging of intact tissues
- 10.1023/b:visi.0000043755.93987.aa may be a valid DOI for title: Computing large deformation metric mappings via geodesic flows of diffeomorphisms
- 10.1146/annurev-bioeng-062117-121105 may be a valid DOI for title: Integrated Biophysical Modeling and Image Analysis: Application to Neuro-Oncology
- 10.1007/978-3-030-59713-9_53 may be a valid DOI for title: Multiatlas calibration of biophysical brain tumor growth models with mass effect
- 10.1137/19m1275280 may be a valid DOI for title: Image-driven biophysical tumor growth model calibration
- 10.1109/tmi.2020.3024264 may be a valid DOI for title: Fully Automatic Calibration of Tumor-Growth Models Using a Single mpMRI Scan
- 10.1007/s11081-018-9390-9 may be a valid DOI for title: PDE-constrained optimization in medical image analysis
- 10.1137/18m1207818 may be a valid DOI for title: CLAIRE: A distributed-memory solver for constrained large deformation diffeomorphic image registration
- 10.1145/3126908.3126930 may be a valid DOI for title: A Framework for Scalable Biophysics-based Image Analysis
- 10.1137/140984002 may be a valid DOI for title: An inexact Newton–Krylov algorithm for constrained diffeomorphic image registration
- 10.1137/16m1070475 may be a valid DOI for title: A Semi-Lagrangian two-level preconditioned Newton–Krylov solver for constrained diffeomorphic image registration
- 10.1109/sc.2016.71 may be a valid DOI for title: Distributed-memory large-deformation diffeomorphic 3D image registration

INVALID DOIs

- None
```

**Comment by mjsottile** [comment]
Hi @andreasmang - while the reviewers start working through their checklists, can you address the missing DOIs in the references?

**Comment by andreasmang** [comment]
> Hi @andreasmang - while the reviewers start working through their checklists, can you address the missing DOIs in the references?

@mjsottile: Thanks. I have added DOIs to all articles for which they exist to the bib file. Changes have been pushed to the repo. Let me know if there are any issues...

**Comment by NMontanaBrown** [comment]
Hi team, I've started review with two specific issues.

[Unit tests](https://github.com/andreasmang/claire/issues/4)
[Contributing Guidelines](https://github.com/andreasmang/claire/issues/5)

**Comment by whedon** [comment]
:wave: @NMontanaBrown, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @tjasaki, please update us on how your review is going (this is an automated reminder).

**Comment by mjsottile** [comment]
:wave: @NMontanaBrown, @tjasaki Let me know if you need any assistance in completing the reviews.  I see that @NMontanaBrown has created some issues in the project that the authors still need to respond so, so I assume that part of the review is ongoing.

**Comment by NMontanaBrown** [comment]
Hi @mjsottile I am waiting for unit tests on the code. I have more questions for the authors but I will wait for this to be addressed before proceeding further.

**Comment by tjasaki** [comment]
Hi @mjsottile I am in the midst of testing the installation.  I am working through missing dependencies and broken links trying to determine the particular causes of the issues.

**Comment by andreasmang** [comment]
@tjasaki can you be more specific. There should not be any missing dependencies.

Considering the other comments. I will post a response soon. I am drowning the last couple of days and I was assuming I get feedback that we address at once as opposed to do this in an iterative way.

**Comment by tjasaki** [comment]
Yes @andreasmang.  I did not mean to imply that the fault could be all at your end.  For example, I needed to install cmake on my system.  Later today I can pass along more particulars about how the build went.  We'll get there.  Tom

**Comment by andreasmang** [comment]
@tjasaki Sorry. I also did not want to imply anything. Feedback is very much appreciated, especially when it comes to compilation related issues. If you want me to update the documentation considering particular issues, do not hesitate. We have several folks involved building this on different systems but they are mostly local (as in I work with them directly) and it's quite complicated due to the different dependencies and the fact that it is developed for HPC architectures...Plan is to increase the visibility and get more folks to use the software. So, again, I am grateful for any hints that help us iron out issues.

**Comment by andreasmang** [comment]
Also, we are currently pushing everything to the GPU. There are less dependencies on that branch, since the FFT kernels are now part of CLAIRE...This will become the default branch soon...

**Comment by andreasmang** [comment]
Dear all:

I apologize for the silence. The last couple of weeks have been pretty intense. I will be able to respond more quickly to further requests from now on.


> Comment R1: Hi @andreasmang, how you have tested/validated your code and its outputs? Currently, as far as I can find, there is no automated test suite via Github Actions or other.

We did not discuss this in the README. We currently do not have support for Github Actions. This remains subject to future work.
CLAIRE features a test binary used in our group to make sure the computational kernels produce accurate results. We have added a description of these to the README file:
https://github.com/andreasmang/claire/blob/gpu/doc/README-RUNME.md#testing

We also provide a brief description of how to use CLAIRE and include a link to data we have extensively considered in our prior work:

https://github.com/andreasmang/nirep

I am not going to mention additional details here but rather refer to the "overhauled" README environment.


> Comment R2: Hi @andreasmang, what are the contributing guidelines for this project? They are missing from the repository.

Response: Thanks for pointing this out. These did not exist. We have added contributing guidelines to the repository:
https://github.com/andreasmang/claire/blob/master/doc/CONTRIBUTING.md

We have also included a "Code of Conduct"
https://github.com/andreasmang/claire/blob/gpu/doc/CODE_OF_CONDUCT.md

Based on a prior comment posted above we have also modified the structure of the README.

We have also enabled "GitHub Discussions" to allow the community to make comments and suggestions.


Cheers A

**Comment by mjsottile** [comment]
:wave: @tjasaki and @NMontanaBrown : it looks like the authors have been making progress in responding to review topics.  Let me know if you need anything to make progress on the reviews.

**Comment by tjasaki** [comment]
@mjsottile : I've been reading and reviewing the new material.  I am still stymied by getting claire to install properly.  I've made progress, it seems to be down to a CUDA (compatibility?) issue which I haven't been able to sort out yet.  The installation instructions seem reasonable and have been notably helpful, but accounting for all possibilities would be unrealistic.

**Comment by andreasmang** [comment]
> @mjsottile : I've been reading and reviewing the new material. I am still stymied by getting claire to install properly. I've made progress, it seems to be down to a CUDA (compatibility?) issue which I haven't been able to sort out yet. The installation instructions seem reasonable and have been notably helpful, but accounting for all possibilities would be unrealistic.

@tjasaki (and others): Let me know if there is anything we should do on our end...(i.e., if you would like us to check if we can emulate a particular setting and if we can figure it out). Feel free to raise an ISSUE or send me an email.

**Comment by mjsottile** [comment]
@andreasmang @tjasaki I’d like to not get blocked on install issues if it boils down to CUDA related problems.  The difficulty in installing tools based on CUDA is often more an issue with how NVidia chooses to distribute their software/drivers and not a reflection of the software that uses it.  I would recommend @tjasaki communicate directly with the CLAIRE developers via GitHub issues or email with the specific problems being encountered on installation, and if the issue can’t be resolved, skip the installation part of the review for now.  I don’t think it’s useful if we get bogged down in third party GPU driver/framework problems that are orthogonal to the package under review.  If everything else looks reasonable on review, we can revisit whether more needs to be done related to installation at the end of the refereeing process.

**Comment by tjasaki** [comment]
@mjsottile: Check.  I've asked for tangential assistance, but will proceed as I am able.

**Comment by mjsottile** [comment]
:wave: @NMontanaBrown and @tjasaki This review appears to have stalled and I want to check to make sure you are both able to continue to make progress.  Please let me or the authors of the submission know if you are blocked and need assistance to move things along.

**Comment by mjsottile** [comment]
@andreasmang apologies for the review stalling out.  I've reached out to the reviewers off of GitHub and if I don't see progress soon I'll be looking for alternative reviewers.

**Comment by tjasaki** [comment]
@mjsottile I spend some time this weekend with the code, its operation, and the review.  I am quite satisfied and have completed my checklist.  It is a neat piece of work.  What is my next step?

**Comment by mjsottile** [comment]
Thanks @tjasaki.  I took a look at the issues opened on the project and saw the exchange related to installation.  Glad it was resolved.  I’ll consider your review complete.  The other reviewer responded over email and has been consumed by other projects, so I’ll take over in their place and do my own review instead of looking for an additional reviewer.  Once I complete my review we’ll be able to move on to the next phase of the paper process.

**Comment by mjsottile** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03038/joss.03038/10.21105.joss.03038.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03038/joss.03038/10.21105.joss.03038.pdf) :page_facing_up: :point_left:

**Comment by mjsottile** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-66182-7_32 is OK
- 10.1038/nprot.2014.123 is OK
- 10.1023/A:1008001603737 is OK
- 10.1007/978-3-642-12055-8 is OK
- 10.1023/B:VISI.0000043755.93987.aa is OK
- 10.1146/annurev-bioeng-062117-121105 is OK
- 10.1109/TMI.2013.2265603 is OK
- 10.1137/1.9780898718843 is OK
- 10.1088/0266-5611/24/3/034008 is OK
- 10.1093/acprof:oso/9780198528418.001.0001 is OK
- 10.1007/978-3-030-59713-9_53 is OK
- 10.1137/19M1275280 is OK
- 10.1109/TMI.2020.3024264 is OK
- 10.1016/j.cma.2018.12.008 is OK
- 10.1007/s11081-018-9390-9 is OK
- 10.1109/SC41405.2020.00042 is OK
- 10.1016/j.jpdc.2020.11.006 is OK
- 10.1137/18M1207818 is OK
- 10.1145/3126908.3126930 is OK
- 10.1137/140984002 is OK
- 10.1137/15m1010919 is OK
- 10.1137/16m1070475 is OK
- 10.1109/sc.2016.71 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mjsottile** [comment]
@whedon remove @NMontanaBrown as reviewer

**Comment by whedon** [comment]
OK, @NMontanaBrown is no longer a reviewer

**Comment by mjsottile** [comment]
@andreasmang I've looked over the submission myself in addition to the review from @tjasaki and I am satisfied that it meets the requirements for JOSS.  The last steps:

- Please create an archive on zenodo or figshare for the submission
- Ensure that the version for the JOSS submission is correct. 
- Ensure that the author list in the paper and archive are correct.

Once you've provided that information I can issue the final commands to pass the submission to an EIC for final processing.

**Comment by andreasmang** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03038/joss.03038/10.21105.joss.03038.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03038/joss.03038/10.21105.joss.03038.pdf) :page_facing_up: :point_left:

**Comment by andreasmang** [comment]
@mjsottile I have looked carefully over the submission. 

- I have fixed the affiliations of my co-authors.
- I have fixed three minor typos in the text and references. 
- I have created an archive on zenodo (https://doi.org/10.5281/zenodo.4837427; I am new to this so let me know if I missed anything; it is a mirror of v0.08 of the code repo; i.e., the current version of the GPU branch). The ZIP file has the version number and ID in the filename. 
- I have pushed a tag to the repo (v0.08) for the JOSS submission (the current version of the code):

Let me know if further action is required on my end and thanks in advance for your help.

**Comment by andreasmang** [comment]
Here's a link that should work: https://zenodo.org/record/4837427

**Comment by mjsottile** [comment]
@whedon set 10.5281/zenodo.4837427 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4837427" target="_blank">10.5281/zenodo.4837427</a> is the archive.

**Comment by mjsottile** [comment]
@whedon set v0.08 as version

**Comment by whedon** [comment]
OK. v0.08 is the version.

**Comment by mjsottile** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-66182-7_32 is OK
- 10.1038/nprot.2014.123 is OK
- 10.1023/A:1008001603737 is OK
- 10.1007/978-3-642-12055-8 is OK
- 10.1023/B:VISI.0000043755.93987.aa is OK
- 10.1146/annurev-bioeng-062117-121105 is OK
- 10.1109/TMI.2013.2265603 is OK
- 10.1137/1.9780898718843 is OK
- 10.1088/0266-5611/24/3/034008 is OK
- 10.1093/acprof:oso/9780198528418.001.0001 is OK
- 10.1007/978-3-030-59713-9_53 is OK
- 10.1137/19M1275280 is OK
- 10.1109/TMI.2020.3024264 is OK
- 10.1016/j.cma.2018.12.008 is OK
- 10.1007/s11081-018-9390-9 is OK
- 10.1109/SC41405.2020.00042 is OK
- 10.1016/j.jpdc.2020.11.006 is OK
- 10.1137/18M1207818 is OK
- 10.1145/3126908.3126930 is OK
- 10.1137/140984002 is OK
- 10.1137/15m1010919 is OK
- 10.1137/16m1070475 is OK
- 10.1109/sc.2016.71 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2346

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2346, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
To recommend a paper to be accepted use `@whedon recommend-accept`

**Comment by arfon** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-66182-7_32 is OK
- 10.1038/nprot.2014.123 is OK
- 10.1023/A:1008001603737 is OK
- 10.1007/978-3-642-12055-8 is OK
- 10.1023/B:VISI.0000043755.93987.aa is OK
- 10.1146/annurev-bioeng-062117-121105 is OK
- 10.1109/TMI.2013.2265603 is OK
- 10.1137/1.9780898718843 is OK
- 10.1088/0266-5611/24/3/034008 is OK
- 10.1093/acprof:oso/9780198528418.001.0001 is OK
- 10.1007/978-3-030-59713-9_53 is OK
- 10.1137/19M1275280 is OK
- 10.1109/TMI.2020.3024264 is OK
- 10.1016/j.cma.2018.12.008 is OK
- 10.1007/s11081-018-9390-9 is OK
- 10.1109/SC41405.2020.00042 is OK
- 10.1016/j.jpdc.2020.11.006 is OK
- 10.1137/18M1207818 is OK
- 10.1145/3126908.3126930 is OK
- 10.1137/140984002 is OK
- 10.1137/15m1010919 is OK
- 10.1137/16m1070475 is OK
- 10.1109/sc.2016.71 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2347

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2347, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1399089906056630275) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2350
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03038](https://doi.org/10.21105/joss.03038)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
@tjasaki – many thanks for your review here and to @mjsottile for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@andreasmang – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03038/status.svg)](https://doi.org/10.21105/joss.03038)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03038">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03038/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03038/status.svg
   :target: https://doi.org/10.21105/joss.03038
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03038/status.svg)](https://doi.org/10.21105/joss.03038)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
